mediatek: filogic: prepare for support of the MT7981 SoC

dangowrt opened the pull request:
Prepare for support of the MediaTek MT7981 SoC and backport pending patches for the Ethernet and DSA driver.
Also dump-down GPY211 driver to behave like a normal PHY driver without assumptions about in-band-status support of the MAC (which aren't true for MediaTek's Ethernet MAC).
Alltogether this prepares the tree for upcoming MT7981 devices, but also improves support for MT7986 and MT7622.

dangowrt committed:
mediatek: backport clk driver for MT7981 SoC

Backport driver for common clocks in MT7981 SoC. The driver has also
been submitted upstream and became part of Linux 6.3.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: backport pinctrl driver for MT7981 SoC

Backport the pinctrl driver for the MT7981 SoC. The driver has also
been submitted upstream and is part of Linux 6.3.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: backport support for MT7981 Ethernet

Backport support for the MT7981 SoC to mtk_eth_soc.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add support for built-in GE PHYs of newer SoCs

MT7981 and the upcoming MT7988 have built-in Gigabit Ethernet PHYs.
While they share some design properties with the PHYs present in
MT753x, they do need calibration data from the SoC's efuse.

Add driver to support them. Upstreaming it is planned, but there are
still some ongoing discussions with MediaTek.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add I2C for MT7981 SoC

Add patch to support I2C on the MT7981 SoC.
This change will also be submitted to upstream Linux soon.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add PWM support for MT7981 SoC

Add patch to support PWM on the MT7981 SoC.
This patch will also be submitted to upstream Linux soon.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: broaden filogic target description

The filogic subtarget now also supports MT7981 and will in future
also support MT7988. Reflect that in the target description.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add support for SPI calibration

Newer MediaTek's SoCs need SPI calibration routines for SPI to work
reliably. Import patches for that from MediaTek's SDK.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: fix Macronix SPI-NAND driver

Use scratch buffer for DMA operetations. Passing a pointer to a stack
variable won't work and results in bogus bit flips being reported.

Patch was submitted upstream and is part of Linux 6.3.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: backport some phylink helper functions

It isn't feasible to literally backport all upstream phylink_pcs changes
down to Linux 5.15: It's just too many patches, and many downstream
drivers and hacks are likely to break. We are too close to branching off
to risk this, and it's also just too much work.
Instead just add helper functions used by modern PCS drivers while keeping
the original functions instact as well. While this may add a kilobyte or
two of extra kernel size, it has the advantage that we get the best of both
worlds: None of the existing codepaths are touched, but yet we have the
option to backport singular improvements to Ethernet drivers where needed.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: simplify MaxLinear GPY PHY driver

The mxl-gpy driver apparently was built in the assumption that SGMII
auto-negotiation is always switched on at the MAC. This may be true for
few rather recent drivers (why?), but certainly isn't for most drivers
unless 'managed = "in-band-status"' is set in device tree. Add patch to
the mediatek target which reduces mxl-gpy to behave more like an
ordinary PHY driver using out-of-band status.

This allows to use these PHYs without rate-adaptation which seems to be
at least partially broken/racy in some revisions of the PHY and/or
internal PHY firmware.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add device tree include for the MT7981 SoC

Add mt7981.dtsi so we can start adding MT7981 devices.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: backport regmap reg_base and downshift features

Backport new regmap features from Linux 5.18 which are needed to let
mt7530 use pcs-mtk-lynxi.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: dsa: make use of pcs-mtk-lynxi also for mt7530

The MT7531 switch IC comes with SerDes ports with PCS identical to
what is also used in MediaTek's SoCs. Make use of the shared driver
to ease maintainance and reduce code duplication.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

ptpt52 commented:
mark.

JiaY-shi commented:
mark

ptpt52 commented:
@dangowrt 
my board build is:
soc: mt7981
switch0: mt7531
switch1: mt7531
lan: 8 x GE ports
wan: 2 x 2.5G ports

gmac0--mt7531--p5--2.5g_phy
gmac1--mt7531--p5--2.5g_phy

dts:
https://github.com/x-wrt/x-wrt/blob/master-mt7981/target/linux/mediatek/dts/mt7981-tenbay_ac-2210e.dts

for now I can run it up, and but 2.5G port not working.

log:
```
## Loading kernel from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'kernel-1' kernel subimage
     Description:  ARM64 OpenWrt Linux-5.15.102
     Type:         Kernel Image
     Compression:  lzma compressed
     Data Start:   0x460000ec
     Data Size:    3869639 Bytes = 3.7 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: 0x48000000
     Entry Point:  0x48000000
     Hash algo:    crc32
     Hash value:   3484f8b6
     Hash algo:    sha1
     Hash value:   5753ea25e061f123454b95b98bffe91345857ae1
   Verifying Hash Integrity ... crc32+ sha1+ OK
## Loading ramdisk from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'initrd-1' ramdisk subimage
     Description:  ARM64 OpenWrt tenbay_ac-2210e initrd
     Type:         RAMDisk Image
     Compression:  Unknown Compression
     Data Start:   0x463b0df0
     Data Size:    12092624 Bytes = 11.5 MiB
     Architecture: AArch64
     OS:           Linux
     Load Address: unavailable
     Entry Point:  unavailable
     Hash algo:    crc32
     Hash value:   973296b6
     Hash algo:    sha1
     Hash value:   2f132c9c607474c7293669db33592c1fbbfa9802
   Verifying Hash Integrity ... crc32+ sha1+ OK
WARNING: 'compression' nodes for ramdisks are deprecated, please fix your .its file!
## Loading fdt from FIT Image at 46000000 ...
   Using 'config-1' configuration
   Trying 'fdt-1' fdt subimage
     Description:  ARM64 OpenWrt tenbay_ac-2210e device tree blob
     Type:         Flat Device Tree
     Compression:  uncompressed
     Data Start:   0x46f393cc
     Data Size:    32768 Bytes = 32 KiB
     Architecture: AArch64
     Hash algo:    crc32
     Hash value:   fca35551
     Hash algo:    sha1
     Hash value:   51088eb6936c869bc6c95b729ee8d5e0fa179e38
   Verifying Hash Integrity ... crc32+ sha1+ OK
   Booting using the fdt blob at 0x46f393cc
   Uncompressing Kernel Image
   Loading Ramdisk to 4ec73000, end 4f7fb4d0 ... OK
   Loading Device Tree to 000000004ec68000, end 000000004ec72fff ... OK

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 5.15.102 (ptpt52@SC-GAME) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 12.2.0 r22297+201-1781e7408a) 12.2.0, GNU ld (GNU Binutils) 2.40.0) #0 SMP Sun Mar 19 04:03:43
[    0.000000] Machine model: Tenbay AC-2210E
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000040000000-0x000000004fffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000040000000-0x0000000042ffffff]
[    0.000000]   node   0: [mem 0x0000000043000000-0x000000004302ffff]
[    0.000000]   node   0: [mem 0x0000000043030000-0x0000000047c7ffff]
[    0.000000]   node   0: [mem 0x0000000047c80000-0x0000000047ffffff]
[    0.000000]   node   0: [mem 0x0000000048000000-0x000000004fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000040000000-0x000000004fffffff]
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.2
[    0.000000] percpu: Embedded 17 pages/cpu s29592 r8192 d31848 u69632
[    0.000000] pcpu-alloc: s29592 r8192 d31848 u69632 alloc=17*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] CPU features: kernel page table isolation disabled by kernel configuration
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64512
[    0.000000] Kernel command line: 
[    0.000000] Dentry cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 5, 131072 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 229248K/262144K available (8128K kernel code, 888K rwdata, 2168K rodata, 384K init, 289K bss, 32896K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu:     RCU restricting CPUs from NR_CPUS=4 to nr_cpu_ids=2.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=2
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 640 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] GICv3: Distributor has no Range Selector support
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: 16 PPIs implemented
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x000000000c080000
[    0.000000] arch_timer: cp15 timer(s) running at 13.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x2ff89eacb, max_idle_ns: 440795202429 ns
[    0.000000] sched_clock: 56 bits at 13MHz, resolution 76ns, wraps every 4398046511101ns
[    0.000148] Calibrating delay loop (skipped), value calculated using timer frequency.. 26.00 BogoMIPS (lpj=130000)
[    0.000158] pid_max: default: 32768 minimum: 301
[    0.000251] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.000262] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.001329] rcu: Hierarchical SRCU implementation.
[    0.001441] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.001697] smp: Bringing up secondary CPUs ...
[    0.002009] Detected VIPT I-cache on CPU1
[    0.002035] GICv3: CPU1: found redistributor 1 region 0:0x000000000c0a0000
[    0.002062] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.002127] smp: Brought up 1 node, 2 CPUs
[    0.002140] SMP: Total of 2 processors activated.
[    0.002145] CPU features: detected: 32-bit EL0 Support
[    0.002149] CPU features: detected: CRC32 instructions
[    0.002235] CPU: All CPU(s) started at EL2
[    0.002247] alternatives: patching kernel code
[    0.002739] devtmpfs: initialized
[    0.005136] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.005158] futex hash table entries: 512 (order: 3, 32768 bytes, linear)
[    0.005251] pinctrl core: initialized pinctrl subsystem
[    0.005812] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.006073] DMA: preallocated 128 KiB GFP_KERNEL pool for atomic allocations
[    0.006101] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.006126] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.006417] thermal_sys: Registered thermal governor 'fair_share'
[    0.006421] thermal_sys: Registered thermal governor 'bang_bang'
[    0.006426] thermal_sys: Registered thermal governor 'step_wise'
[    0.006430] thermal_sys: Registered thermal governor 'user_space'
[    0.006581] ASID allocator initialised with 65536 entries
[    0.015877] cryptd: max_cpu_qlen set to 1000
[    0.017965] SCSI subsystem initialized
[    0.018125] libata version 3.00 loaded.
[    0.019091] clocksource: Switched to clocksource arch_sys_counter
[    0.019776] NET: Registered PF_INET protocol family
[    0.019871] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.020168] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.020189] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.020199] TCP established hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.020218] TCP bind hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.020246] TCP: Hash tables configured (established 2048 bind 2048)
[    0.020305] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.020322] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.020423] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.020445] PCI: CLS 0 bytes, default 64
[    0.020699] Unpacking initramfs...
[    0.022619] workingset: timestamp_bits=46 max_order=16 bucket_order=0
[    0.025399] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.025415] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.069693] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.071199] printk: console [ttyS0] disabled
[    0.091362] 11002000.serial: ttyS0 at MMIO 0x11002000 (irq = 18, base_baud = 2500000) is a ST16650V2
[    0.695113] printk: console [ttyS0] enabled
[    0.700618] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.710580] loop: module loaded
[    0.714250] mt7981-pinctrl 11d00000.pinctrl: pin_config_set op failed for pin 19
[    0.721672] mtk-spi 1100a000.spi: Error applying setting, reverse things back
[    0.731007] spi-nand spi0.0: calibration result: 0x3
[    0.736094] spi-nand spi0.0: Winbond SPI NAND was found.
[    0.741445] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.753488] [BBT] BMT.v2 is found at 0x3ff
[    3.603480] Freeing initrd memory: 11808K
[    3.620459] 8 fixed-partitions partitions found on MTD device spi0.0
[    3.627255] Creating 8 MTD partitions on "spi0.0":
[    3.632067] 0x000000000000-0x000000100000 : "BL2"
[    3.637927] 0x000000100000-0x000000180000 : "u-boot-env"
[    3.644153] 0x000000180000-0x000000380000 : "Factory"
[    3.651107] 0x000000380000-0x000000580000 : "FIP"
[    3.657499] 0x000000580000-0x000003580000 : "ubi"
[    3.692997] 0x000003580000-0x000006580000 : "ubi1"
[    3.728552] 0x000006580000-0x0000065a0000 : "Product"
[    3.734195] 0x0000065a0000-0x0000075a0000 : "Custom"
[    3.909971] MediaTek MT7981 PHY mdio-bus:00: TX-VCM SW cal result: 0x2
[    3.949467] mtk_soc_eth 15100000.ethernet eth0: mediatek frame engine at 0xffffffc009180000, irq 82
[    3.959250] mtk_soc_eth 15100000.ethernet: generated random MAC address 2a:31:c5:63:d7:5d
[    3.968110] mtk_soc_eth 15100000.ethernet eth1: mediatek frame engine at 0xffffffc009180000, irq 82
[    3.977813] i2c_dev: i2c /dev entries driver
[    3.983533] mtk-wdt 1001c000.watchdog: Watchdog enabled (timeout=31 sec, nowayout=0)
[    3.992148] NET: Registered PF_INET6 protocol family
[    3.997904] Segment Routing with IPv6
[    4.001616] In-situ OAM (IOAM) with IPv6
[    4.005573] NET: Registered PF_PACKET protocol family
[    4.010670] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    4.023757] 8021q: 802.1Q VLAN Support v1.8
[    4.571200] mt7530 mdio-bus:1f: no interrupt support
[    4.642153] mt7530 mdio-bus:1f: configuring for fixed/2500base-x link mode
[    4.651095] mt7530 mdio-bus:1f: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    4.660043] mt7530 mdio-bus:1f lan1 (uninitialized): PHY [mt7530-0:00] driver [MediaTek MT7531 PHY] (irq=POLL)
[    4.680109] mt7530 mdio-bus:1f lan2 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7531 PHY] (irq=POLL)
[    4.699876] mt7530 mdio-bus:1f lan3 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7531 PHY] (irq=POLL)
[    4.719620] mt7530 mdio-bus:1f lan4 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7531 PHY] (irq=POLL)
[    4.739372] mt7530 mdio-bus:1f gsw1_p4 (uninitialized): PHY [mt7530-0:04] driver [MediaTek MT7531 PHY] (irq=POLL)
[    4.799104] mt7530 mdio-bus:1f wan1 (uninitialized): PHY [mdio-bus:07] driver [RTL8221B-VB-CG 2.5Gbps PHY] (irq=POLL)
[    4.810196] DSA: tree 0 setup
[    4.814052] UBI: auto-attach mtd4
[    4.817378] ubi0: default fastmap pool size: 15
[    4.821911] ubi0: default fastmap WL pool size: 7
[    4.826602] ubi0: attaching mtd4
[    5.142611] ubi0: scanning is finished
[    5.155535] ubi0 warning: ubi_eba_init: cannot reserve enough PEBs for bad PEB handling, reserved 17, need 19
[    5.166130] ubi0: attached mtd4 (name "ubi", size 48 MiB)
[    5.171551] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    5.178414] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    5.185199] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    5.192151] ubi0: good PEBs: 384, bad PEBs: 0, corrupted PEBs: 0
[    5.198144] ubi0: user volume: 3, internal volumes: 1, max. volumes count: 128
[    5.205357] ubi0: max/mean erase counter: 7/4, WL threshold: 4096, image sequence number: 1292923203
[    5.214481] ubi0: available PEBs: 0, total reserved PEBs: 384, PEBs reserved for bad PEB handling: 17
[    5.223695] ubi0: background thread "ubi_bgt0d" started, PID 434
[    5.231628] block ubiblock0_1: created from ubi0:1(rootfs)
[    5.237118] ubiblock: device ubiblock0_1 (rootfs) set to be root filesystem
[    5.244860] Freeing unused kernel memory: 384K
[    5.289114] Run /init as init process
[    5.292766]   with arguments:
[    5.295720]     /init
[    5.297981]   with environment:
[    5.301116]     HOME=/
[    5.303463]     TERM=linux
[    5.838898] init: Console is alive
[    5.842970] init: - watchdog -
[    5.855969] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    5.866881] usbcore: registered new interface driver usbfs
[    5.872433] usbcore: registered new interface driver hub
[    5.877787] usbcore: registered new device driver usb
[    5.883320] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    5.896854] init: - preinit -
[    5.948540] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[    5.957009] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
[    5.966529] mt7530 mdio-bus:1f lan1: configuring for phy/gmii link mode
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[    8.020920] procd: - early -
[    8.023872] procd: - watchdog -
[    8.550204] procd: - watchdog -
[    8.553512] procd: - ubus -
[    8.558695] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.606012] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.612546] random: ubusd: uninitialized urandom read (4 bytes read)
[    8.619862] procd: - init -
Please press Enter to activate this console.
[    8.744485] urngd: v1.0.2 started.
[    8.771355] random: crng init done
[    8.774761] random: 21 urandom warning(s) missed due to ratelimiting
[    8.791459] kmodloader: loading kernel modules from /etc/modules.d/*
[    8.807872] tun: Universal TUN/TAP device driver, 1.6
[    8.818187] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    8.828361] gre: GRE over IPv4 demultiplexor driver
[    8.834603] ip_gre: GRE over IPv4 tunneling driver
[    8.841907] ip6_gre: GRE over IPv6 tunneling driver
[    8.882249] i2c-gpio i2c@0: using lines 477 (SDA) and 478 (SCL)
[    8.891274] i2c-gpio i2c@1: using lines 479 (SDA) and 480 (SCL)
[    8.899009] GACT probability on
[    8.903400] Mirror/redirect action on
[    8.911727] u32 classifier
[    8.914438]     input device check on
[    8.918091]     Actions configured
[    8.927122] crypto-safexcel 10320000.crypto: EIP97:230(0,1,4,4)-HIA:270(0,5,5),PE:150/433(alg:7fcdfc00)/0/0/0
[    8.952215] {natcap}:natcap_init(): version: 8.0.0
[    8.957012] {natcap}:natcap_init(): natcap_major=249, natcap_minor=0
[    8.965033] {natcap}:natcap_peer_init(): natcap_peer_major=248, natcap_peer_minor=0
[    8.975075] {natflow}:natflow_init(): natflow_major=247, natflow_minor=0
[    8.982047] {natflow}:natflow_zone_init(): natflow_zone_major=246, natflow_zone_minor=0
[    8.990335] {natflow}:natflow_user_init(): natflow_user_major=245, natflow_user_minor=0
[    8.998514] {natflow}:userinfo_init(): userinfo_major=244, userinfo_minor=0
[    9.006420] {natflow}:userinfo_event_init(): userinfo_event_major=243, userinfo_event_minor=0
[    9.015601] {natflow}:qos_init(): qos_major=242, qos_minor=0
[    9.023140] {natflow}:conntrackinfo_init(): conntrackinfo_major=241, conntrackinfo_minor=0
[    9.031710] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=eth0(tso=1,1,hw_csum=1,1), add ingress hook
[    9.042563] {natflow}:natflow_urllogger_init(): urllogger_major=240, urllogger_minor=0
[    9.056707] Bridge firewalling registered
[    9.061674] Loading modules backported from Linux version v6.1-rc8-0-g76dcd734eca2
[    9.069276] Backport generated by backports.git v5.15.81-1-41-g02e352527db5
[   10.021820] mt7530 gpio-0:1f: no interrupt support
[   10.238019] mt7530 gpio-0:1f: configuring for fixed/2500base-x link mode
[   10.287085] mt7530 gpio-0:1f: Link is Up - 2.5Gbps/Full - flow control rx/tx
[   10.487491] mt7530 gpio-0:1f lan5 (uninitialized): PHY [mt7530-1:00] driver [MediaTek MT7531 PHY] (irq=POLL)
[   10.707983] mt7530 gpio-0:1f lan6 (uninitialized): PHY [mt7530-1:01] driver [MediaTek MT7531 PHY] (irq=POLL)
[   10.924628] mt7530 gpio-0:1f lan7 (uninitialized): PHY [mt7530-1:02] driver [MediaTek MT7531 PHY] (irq=POLL)
[   11.141255] mt7530 gpio-0:1f lan8 (uninitialized): PHY [mt7530-1:03] driver [MediaTek MT7531 PHY] (irq=POLL)
[   11.357864] mt7530 gpio-0:1f gsw2_p4 (uninitialized): PHY [mt7530-1:04] driver [MediaTek MT7531 PHY] (irq=POLL)
[   11.439150] mt7530 gpio-0:1f wan2 (uninitialized): PHY [gpio-0:07] driver [RTL8221B-VB-CG 2.5Gbps PHY] (irq=POLL)
[   11.449950] DSA: tree 1 setup
[   11.534358] xt_time: kernel timezone is -0000
[   11.578524] PPP generic driver version 2.4.2
[   11.583613] PPP MPPE Compression module registered
[   11.589291] NET: Registered PF_PPPOX protocol family
[   11.595101] PPTP driver version 0.8.5
[   11.600795] wireguard: WireGuard 1.0.0 loaded. See www.wireguard.com for information.
[   11.608618] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[   11.620128] kmodloader: done loading kernel modules from /etc/modules.d/*
[   14.148068] mtdblock: MTD device 'BL2' is NAND, please consider using UBI block devices instead.
[   14.157007] mtdblock: MTD device 'BL2' is NAND, please consider using UBI block devices instead.
[   14.159181] /dev/sda1: Can't open blockdev
[   14.173847] mtdblock: MTD device 'u-boot-env' is NAND, please consider using UBI block devices instead.
[   14.183331] mtdblock: MTD device 'u-boot-env' is NAND, please consider using UBI block devices instead.
[   14.201022] mtdblock: MTD device 'Factory' is NAND, please consider using UBI block devices instead.
[   14.210289] mtdblock: MTD device 'Factory' is NAND, please consider using UBI block devices instead.
[   14.227496] mtdblock: MTD device 'FIP' is NAND, please consider using UBI block devices instead.
[   14.236415] mtdblock: MTD device 'FIP' is NAND, please consider using UBI block devices instead.
[   14.253376] mtdblock: MTD device 'ubi' is NAND, please consider using UBI block devices instead.
[   14.262372] mtdblock: MTD device 'ubi' is NAND, please consider using UBI block devices instead.
[   14.286498] mtdblock: MTD device 'ubi1' is NAND, please consider using UBI block devices instead.
[   14.295951] mtdblock: MTD device 'ubi1' is NAND, please consider using UBI block devices instead.
[   14.321002] mtdblock: MTD device 'Product' is NAND, please consider using UBI block devices instead.
[   14.330275] mtdblock: MTD device 'Product' is NAND, please consider using UBI block devices instead.
[   14.359241] mtdblock: MTD device 'Custom' is NAND, please consider using UBI block devices instead.
[   14.368342] mtdblock: MTD device 'Custom' is NAND, please consider using UBI block devices instead.
[   19.505537] mtk_soc_eth 15100000.ethernet eth0: Link is Down
[   19.518389] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/2500base-x link mode
[   19.528275] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=eth0(tso=1,1,hw_csum=1,1), add ingress hook
[   19.528474] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 2.5Gbps/Full - flow control rx/tx
[   19.547740] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   19.560326] mt7530 mdio-bus:1f lan1: configuring for phy/gmii link mode
[   19.570105] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan1(tso=1,1,hw_csum=1,1), add ingress hook
[   19.588118] br-lan: port 1(lan1) entered blocking state
[   19.593436] br-lan: port 1(lan1) entered disabled state
[   19.601488] device lan1 entered promiscuous mode
[   19.606109] device eth0 entered promiscuous mode
[   19.624544] mt7530 mdio-bus:1f lan2: configuring for phy/gmii link mode
[   19.636764] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan2(tso=1,1,hw_csum=1,1), add ingress hook
[   19.651202] br-lan: port 2(lan2) entered blocking state
[   19.656442] br-lan: port 2(lan2) entered disabled state
[   19.664263] device lan2 entered promiscuous mode
[   19.675477] mt7530 mdio-bus:1f lan3: configuring for phy/gmii link mode
[   19.682716] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan3(tso=1,1,hw_csum=1,1), add ingress hook
[   19.696184] br-lan: port 3(lan3) entered blocking state
[   19.697892] mt7530 mdio-bus:1f lan3: Link is Up - 1Gbps/Full - flow control rx/tx
[   19.701460] br-lan: port 3(lan3) entered disabled state
[   19.720322] device lan3 entered promiscuous mode
[   19.727208] br-lan: port 3(lan3) entered blocking state
[   19.732471] br-lan: port 3(lan3) entered forwarding state
[   19.740198] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   19.752447] mt7530 mdio-bus:1f lan4: configuring for phy/gmii link mode
[   19.759953] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan4(tso=1,1,hw_csum=1,1), add ingress hook
[   19.780924] br-lan: port 4(lan4) entered blocking state
[   19.786163] br-lan: port 4(lan4) entered disabled state
[   19.808585] device lan4 entered promiscuous mode
[   19.820946] mt7530 mdio-bus:1f wan1: configuring for phy/sgmii link mode
[   19.869317] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=wan1(tso=1,1,hw_csum=1,1), add ingress hook
[   19.887363] br-lan: port 5(wan1) entered blocking state
[   19.892659] br-lan: port 5(wan1) entered disabled state
[   19.903565] device wan1 entered promiscuous mode
[   19.914380] mtk_soc_eth 15100000.ethernet eth1: configuring for fixed/2500base-x link mode
[   19.926807] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=eth1(tso=1,1,hw_csum=1,1), add ingress hook
[   19.937696] mtk_soc_eth 15100000.ethernet eth1: Link is Up - 2.5Gbps/Full - flow control rx/tx
[   19.957831] mt7530 gpio-0:1f lan5: configuring for phy/gmii link mode
[   19.974035] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan5(tso=1,1,hw_csum=1,1), add ingress hook
[   20.025843] br-lan: port 6(lan5) entered blocking state
[   20.031110] br-lan: port 6(lan5) entered disabled state
[   20.066267] device lan5 entered promiscuous mode
[   20.070929] device eth1 entered promiscuous mode
[   20.118189] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[   20.180211] mt7530 gpio-0:1f lan6: configuring for phy/gmii link mode
[   20.196379] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan6(tso=1,1,hw_csum=1,1), add ingress hook
[   20.252725] br-lan: port 7(lan6) entered blocking state
[   20.257960] br-lan: port 7(lan6) entered disabled state
[   20.296240] device lan6 entered promiscuous mode
[   20.391746] mt7530 gpio-0:1f lan7: configuring for phy/gmii link mode
[   20.407919] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan7(tso=1,1,hw_csum=1,1), add ingress hook
[   20.461146] br-lan: port 8(lan7) entered blocking state
[   20.466385] br-lan: port 8(lan7) entered disabled state
[   20.508518] device lan7 entered promiscuous mode
[   20.576922] mt7530 gpio-0:1f lan8: configuring for phy/gmii link mode
[   20.593140] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=lan8(tso=1,1,hw_csum=1,1), add ingress hook
[   20.656525] br-lan: port 9(lan8) entered blocking state
[   20.661822] br-lan: port 9(lan8) entered disabled state
[   20.707788] device lan8 entered promiscuous mode
[   20.779182] mt7530 gpio-0:1f wan2: configuring for phy/sgmii link mode
[   20.849321] {natflow}:natflow_netdev_event(): catch NETDEV_UP event for dev=wan2(tso=1,1,hw_csum=1,1), add ingress hook
```

ptpt52 commented:
Two 2.5G phys can be detected. The 2.5G phy on the first 7531 can link up, but the network interface is not functional. The 2.5G phy on the second 7531, which is on the gpio simulated mdio bus, cannot link up.

aiamadeus commented on target/linux/generic/pending-5.15/742-net-dsa-mt7530-use-external-PCS-driver.patch:
```diff
@@ -0,0 +1,514 @@
+ 	case PHY_INTERFACE_MODE_1000BASEX:
+ 	case PHY_INTERFACE_MODE_2500BASEX:
+-		return mt7531_sgmii_setup_mode_force(priv, port, interface);
++		/* handled in SGMII PCS driver */
```
Sorry but there seems to be a problem here? I can't connect with 1g speed on mt7531 + rtl8221b.

dangowrt replied:
@bmork Has been trying this on devices having a 2.5G PHY connected to MT7531, I think it was also RTL8221B, so maybe he can help there. I only got a board with GPY211.

dangowrt replied:
I've been reading `drivers/net/phy/realtek.c` and there is nothing there switching the interface mode between SGMII and 2500Base-X. So this phy driver will need more work to inform the system about this, similar to how it is done in `mxl-gpy.c`, there you can see function `gpy_update_interface` taking care of this.

aiamadeus replied:
Of course I applied my patch first, and `ethtool lan2` looks fine:
![t](https://user-images.githubusercontent.com/42570690/226178960-15b5dda4-c1cd-4aad-8b19-167448586d2c.png)

dangowrt replied:
Please dump SerDes related registers of the MT7531:
```
cat /sys/kernel/debug/regmap/mdio-bus:1f-port5/registers
```

aiamadeus replied:
https://pastebin.com/3naXXqsG

dangowrt replied:
Please also check the value of this register of the RealTek PHY.
![serdes-an](https://user-images.githubusercontent.com/9948313/226182209-398f0ba8-2bee-40be-8034-1882a6275789.png)

If SerDes AN is enabled, SerDes link will not come up unless it is enabled also on the other side (ie. MT7531). As SerDes AN (aka. in-band-status) is typically not use for MAC<->PHY connections but mostly for SFP links, you are advised to make sure it is switched off also on the RealTek PHY.

To play with it you can use `mdio-tools`

dangowrt replied:
> https://pastebin.com/3naXXqsG
> ```
> 000: 00040140
> ```

This looks like SerDes link is up from perspective of the MT7531 (bit 18 is 1). However, as expected when in-band-status is not in use SGMII autonegotiation is switched off (bit 12 is 0). Hence the same needs to be true on the side of the PHY, ie. MMD device 30, register 0x7582 bit 7 needs to be 0.

aiamadeus replied:
> ie. MMD device 30, register 0x7582 bit 7 needs to be 0.

Thanks, am I doing this correctly?
Read: `mdio mdio-bus mmd 5:30 raw 0x7582` Output: 0x49f6
Write: `mdio mdio-bus mmd 5:30 raw 0x7582 0x4976`

dangowrt replied:
Yes, looks correct.
And indeed bit 7 was originally set (maybe by the bootloader or post-reset default value). Make sure to trigger a re-connect (ie. pull and replug cable, or even plug a 2500Base-T link and then try again with 1000Base-T, so the PHY realizes that it should re-try establishing a link to the MAC.

dangowrt replied:
You can try this patch
```diff
diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index 3d99fd6664d7a..3b1f32a471961 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -675,6 +675,17 @@ static int rtl822x_config_aneg(struct phy_device *phydev)
        return __genphy_config_aneg(phydev, ret);
 }
 
+static int rtl8221_config_aneg(struct phy_device *phydev)
+{
+       int ret = 0;
+
+       ret = phy_modify_mmd_changed(phydev, MDIO_MMD_VEND1, 0x7582, BIT(7), 0);
+       if (ret == 1)
+               ret = genphy_soft_reset(phydev);
+
+       return rtl822x_config_aneg(phydev);
+}
+
 static int rtl822x_read_status(struct phy_device *phydev)
 {
        int ret;
@@ -1001,7 +1012,7 @@ static struct phy_driver realtek_drvs[] = {
                PHY_ID_MATCH_EXACT(0x001cc849),
                .name           = "RTL8221B-VB-CG 2.5Gbps PHY",
                .get_features   = rtl822x_get_features,
-               .config_aneg    = rtl822x_config_aneg,
+               .config_aneg    = rtl8221_config_aneg,
                .read_status    = rtl822x_read_status,
                .suspend        = genphy_suspend,
                .resume         = rtlgen_resume,
@@ -1011,7 +1022,7 @@ static struct phy_driver realtek_drvs[] = {
                PHY_ID_MATCH_EXACT(0x001cc84a),
                .name           = "RTL8221B-VM-CG 2.5Gbps PHY",
                .get_features   = rtl822x_get_features,
-               .config_aneg    = rtl822x_config_aneg,
+               .config_aneg    = rtl8221_config_aneg,
                .read_status    = rtl822x_read_status,
                .suspend        = genphy_suspend,
                .resume         = rtlgen_resume,
```

Also let me know which exact PHY you are trying this with.

aiamadeus replied:
> You can try this patch [PHY_ID(0x001cc849)]

Unfortunately neither using mdio-tools nor this patch can make it work.

dangowrt replied:
Please share a complete bootlog and also the exact type of the connected PHY.

dangowrt replied:
I assume you are using a RTL8221 variant.
On top of https://github.com/openwrt/openwrt/commit/674ee0bacde4b40ad7bea29698daed3ad242dc57, please additionally apply this patch to make sure the PHY is set to correct 2500Base-X+SGMII mode.

```diff
diff --git a/drivers/net/phy/realtek.c b/drivers/net/phy/realtek.c
index 3d99fd6664d7a..33cbd0f2fe4c3 100644
--- a/drivers/net/phy/realtek.c
+++ b/drivers/net/phy/realtek.c
@@ -53,6 +53,26 @@
 						 RTL8201F_ISR_LINK)
 #define RTL8201F_IER				0x13
 
+#define RTL8221_SERDES_OPTION_1			0x697A
+#define RTL8221_SERDES_MODE_MASK		GENMASK(5, 0)
+#define RTL8221_SERDES_MODE_2500BASEX_SGMII	0x0
+#define RTL8221_SERDES_MODE_HISGMII_SGMII	0x1
+#define RTL8221_SERDES_MODE_2500BASEX		0x2
+#define RTL8221_SERDES_MODE_HISGMII		0x3
+#define RTL8221_SERDES_LINK_DOWN_OPTION		BIT(15)
+
+#define RTL8221_SERDES_CONTROL_5		0x7582
+#define RTL8221_FULL_DUPLEX			BIT(2)
+#define RTL8221_RX_PAUSE			BIT(5)
+#define RTL8221_TX_PAUSE			BIT(6)
+#define RTL8221_AN_ENABLE			BIT(7)
+#define RTL8221_PORT_SPEED_MASK			(GENMASK(13, 12) | GENMASK(1, 0))
+#define RTL8221_PORT_SPEED_10M			0x0
+#define RTL8221_PORT_SPEED_100M			0x1
+#define RTL8221_PORT_SPEED_1000M		0x2
+#define RTL8221_PORT_SPEED_2500M		0x1001
+#define RTL8221_PORT_SPEED_2500M_LITE		0x1003
+
 #define RTL8366RB_POWER_SAVE			0x15
 #define RTL8366RB_POWER_SAVE_ON			BIT(12)
 
@@ -675,6 +695,22 @@ static int rtl822x_config_aneg(struct phy_device *phydev)
 	return __genphy_config_aneg(phydev, ret);
 }
 
+static int rtl8221_config_init(struct phy_device *phydev)
+{
+	/* set 2500Base-X+SGMII mode */
+	phy_modify_mmd(phydev, MDIO_MMD_VEND1, RTL8221_SERDES_OPTION_1,
+		       RTL8221_SERDES_MODE_MASK |
+		       RTL8221_SERDES_LINK_DOWN_OPTION,
+		       RTL8221_SERDES_MODE_2500BASEX_SGMII |
+		       RTL8221_SERDES_LINK_DOWN_OPTION);
+
+	/* disable SerDes AN */
+	phy_clear_bits_mmd(phydev, MDIO_MMD_VEND1,
+			   RTL8221_SERDES_CONTROL_5, RTL8221_AN_ENABLE);
+
+	return 0;
+}
+
 static int rtl822x_read_status(struct phy_device *phydev)
 {
 	int ret;
@@ -1002,6 +1038,7 @@ static struct phy_driver realtek_drvs[] = {
 		.name           = "RTL8221B-VB-CG 2.5Gbps PHY",
 		.get_features   = rtl822x_get_features,
 		.config_aneg    = rtl822x_config_aneg,
+		.config_init    = rtl8221_config_init,
 		.read_status    = rtl822x_read_status,
 		.suspend        = genphy_suspend,
 		.resume         = rtlgen_resume,
@@ -1012,6 +1049,7 @@ static struct phy_driver realtek_drvs[] = {
 		.name           = "RTL8221B-VM-CG 2.5Gbps PHY",
 		.get_features   = rtl822x_get_features,
 		.config_aneg    = rtl822x_config_aneg,
+		.config_init    = rtl8221_config_init,
 		.read_status    = rtl822x_read_status,
 		.suspend        = genphy_suspend,
 		.resume         = rtlgen_resume,
```

I cannot test because I don't have any device using this PHY at hand, but you get the idea. Also I don't know if we really need to set `RTL8221_SERDES_LINK_DOWN_OPTION`, the datasheet is a bit ambiguous when it comes to that bit, so you can try with setting it to 1 and to 0 and see if that helps.

Generally we are interested in registers
 - 30.0x697A SERDES OPTION
 - 30.0x6A04 SERDES Control 1 Register
 - 30.0x7580 SERDES Control 3 Register
 - 30.0x7581 SERDES Control 4 Register
 - 30.0x7582 SERDES Control 5 Register
 - 30.0x7587 SERDES Control 6 Register
 - 30.0x758B SERDES Control 7 Register
 - 30.0x758C SERDES Control 8 Register
 - 30.0x758D SERDES Control 9 Register

So in case the patch above doesn't help to bring the link up, please dump those registers using `mdio` tool.

aiamadeus replied:
[RTL8221B-VB-CG (0x001cc849)]
https://pastebin.com/LhUtr5Qb

aiamadeus replied:
On the current source of openwrt master, the rtl8221b phy connected to mt7531 can work normally (with my patch). It was broken after applying the changes in this pull request, and the patch above doesn't fix it. Sorry I don't have more time to test it.

ptpt52 replied:
@dangowrt  
I have a board consisting of 7981, two MT7531 chips, and two RTL8221B_PHY chips. I believe I am capable of conducting the test, but unfortunately, the RTL8221B_PHY chips are not currently functioning. Do you have any suggestions for troubleshooting the issue?

dangowrt replied:
@ptpt52 I believe the problem is related to the RTL8221B PHYs not being initialized correctly. Apart from the patch by @aiamadeus we will need to introduce a `.config_init` function which sets the PHY to operate in SGMII mode without in-band-status for 10M/100M/1000M and switch to 2500Base-X for 2500M. I've (blindly because I don't have the hardware yet) created such a `.config_init` function, but maybe other things are still wrong there. However, it should provide a good starting point:
https://github.com/openwrt/openwrt/pull/12202#discussion_r1141632220

As @aiamadeus has already reported that the patch doesn't work as-is, my next try would be to remove `RTL8221_SERDES_LINK_DOWN_OPTION` and re-try without that option set. The datasheets you can find also mention that there is a "SerDes application notes" document which I don't have, and probably that document mentions the correct sequence how to initialize the PHY for any possible modes....

dangowrt commented:
> The 2.5G phy on the second 7531, which is on the gpio simulated mdio bus, cannot link up.

Polling 6 PHYs via GPIO bit-banging MDIO will create significant CPU load. If you know the board well, maybe interrupt for MT7531 is connected? Using it would significantly reduce the load on the bit-banging MDIO...

mrkiko commented:
Nice work. Curious to try this out.

blocktrron commented:
Thanks for the series. This fixes the 2.5GE WAN Port on the TUF-AX4200 which previously only worked with 2.5G physical links.

ptpt52 commented:
build failed.
```
make[5]: Entering directory '/mnt/Data/Sources/openwrt/x-wrt/tmpfs/target-aarch64_cortex-a53_musl/linux-mediatek_filogic/linux-5.15.104'
  CALL    scripts/checksyscalls.sh
  CALL    scripts/atomic/check-atomics.sh
  CHK     include/generated/compile.h
  CC      drivers/net/ethernet/mediatek/mtk_eth_soc.o
In file included from drivers/net/ethernet/mediatek/mtk_eth_soc.c:29:
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1269:49: warning: 'struct mtk_sgmii' declared inside parameter list will not be visible outside of this definition or declaration
 1269 | struct phylink_pcs *mtk_sgmii_select_pcs(struct mtk_sgmii *ss, int id);
      |                                                 ^~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1270:27: warning: 'struct mtk_sgmii' declared inside parameter list will not be visible outside of this definition or declaration
 1270 | int mtk_sgmii_init(struct mtk_sgmii *ss, struct device_node *np,
      |                           ^~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.c:4449:12: error: conflicting types for 'mtk_sgmii_init'; have 'int(struct mtk_eth *)'
 4449 | static int mtk_sgmii_init(struct mtk_eth *eth)
      |            ^~~~~~~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1270:5: note: previous declaration of 'mtk_sgmii_init' with type 'int(struct mtk_sgmii *, struct device_node *, u32)' {aka 'int(struct mtk_sgmii *, struct device_node *, unsigned int)'}
 1270 | int mtk_sgmii_init(struct mtk_sgmii *ss, struct device_node *np,
      |     ^~~~~~~~~~~~~~
scripts/Makefile.build:289: recipe for target 'drivers/net/ethernet/mediatek/mtk_eth_soc.o' failed
```

ptpt52 commented:
build failed.
```
make[5]: Entering directory '/mnt/Data/Sources/openwrt/x-wrt/tmpfs/target-aarch64_cortex-a53_musl/linux-mediatek_filogic/linux-5.15.104'
  CALL    scripts/checksyscalls.sh
  CALL    scripts/atomic/check-atomics.sh
  CHK     include/generated/compile.h
  CC      drivers/net/ethernet/mediatek/mtk_eth_soc.o
In file included from drivers/net/ethernet/mediatek/mtk_eth_soc.c:29:
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1269:49: warning: 'struct mtk_sgmii' declared inside parameter list will not be visible outside of this definition or declaration
 1269 | struct phylink_pcs *mtk_sgmii_select_pcs(struct mtk_sgmii *ss, int id);
      |                                                 ^~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1270:27: warning: 'struct mtk_sgmii' declared inside parameter list will not be visible outside of this definition or declaration
 1270 | int mtk_sgmii_init(struct mtk_sgmii *ss, struct device_node *np,
      |                           ^~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.c:4449:12: error: conflicting types for 'mtk_sgmii_init'; have 'int(struct mtk_eth *)'
 4449 | static int mtk_sgmii_init(struct mtk_eth *eth)
      |            ^~~~~~~~~~~~~~
drivers/net/ethernet/mediatek/mtk_eth_soc.h:1270:5: note: previous declaration of 'mtk_sgmii_init' with type 'int(struct mtk_sgmii *, struct device_node *, u32)' {aka 'int(struct mtk_sgmii *, struct device_node *, unsigned int)'}
 1270 | int mtk_sgmii_init(struct mtk_sgmii *ss, struct device_node *np,
      |     ^~~~~~~~~~~~~~
scripts/Makefile.build:289: recipe for target 'drivers/net/ethernet/mediatek/mtk_eth_soc.o' failed
```

dangowrt committed:
generic: replace mtk_sgmii patches with versions accepted upstream

Replace patches for MediaTek Ethernet driver SGMII/SerDes unit with
their corresponding upstream patches. Not all of the patches in our
tree went upstream as-is, some are slightly different implementations,
and they require the phylink_pcs helpers now made available.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: mtk_eth_soc: allow setting MDIO clock frequency

Backport patch allowing to set the MDIO bus clock frequency.
By default the MDIO bus clock runs on 2.5 MHz, allow increasing it
up to 25 MHz.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: net: pcs: add driver for MediaTek SGMII PCS

Backport dedicated PCS driver for MediaTek LynxI SGMII/SerDes unit.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: mtk_eth_soc: switch to external PCS driver

Backport patch to make use of the new PCS driver in mtk_eth_soc.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
generic: pcs-mtk-lynxi: add hack to use 2500Base-X without AN

Using 2500Base-T SFP modules e.g. on the BananaPi R3 requires manually
disabling auto-negotiation, e.g. using ethtool. While a proper fix
using SFP quirks is being discussed upstream, bring a work-around to
restore user experience to what it was before the switch to the
dedicated SGMII PCS driver.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

ptpt52 commented:
@dangowrt 
there is still issue on my board (mt7981 + mt7531 x 2 + 2.5G phy x 2)
where 2.5G phy connect to mt7531 port5
1. first 2.5G phy on mt7531 link detected yes, but cannot send/recv data.
2. sencond 2.5G phy on mt7531_2 could not even detected link.

ptpt52 commented:
@dangowrt 
Thank you. 
I wanted to provide some additional information regarding the issue you mentioned with the second PHY and MT7531. We have found that they are functioning properly using the MTK SDK code. However, we have encountered issues when using the latest OpenWrt master code.

Additionally, I wanted to let you know that we are actively working to improve this board and it is still in the development phase. I appreciate your suggestion to connect the interrupt pins of the second PHY and MT7531 to specific GPIOs, and we will take this into consideration as we continue to make improvements.

ptpt52 commented:
@dangowrt 
>Do you have GPL sources for this board which could serve as a reference?

Maybe here the code for this board
https://github.com/x-wrt/x-wrt/blob/master/target/linux/mediatek/dts/mt7981-tenbay_ac-2210e.dts
or https://github.com/x-wrt/x-wrt

blocktrron commented:
@dangowrt Thanks for your effort.

dangowrt commented:
> or https://github.com/x-wrt/x-wrt

I can see that you have a patch from @lynxis which introduces `.config_init` function like I thought to the PHY driver.
Did you try using this patch with OpenWrt as well?
https://github.com/x-wrt/x-wrt/blob/master/target/linux/generic/pending-5.15/721-net-phy-realtek-rtl8221-allow-to-configure-SERDES-mo.patch

bmork commented:
> I've ordered a device which combines MT7986 and RTL8221BV in the hope to learn more about the PHY and fix the issue once it arrives (~ mid April).

Don't know if you're aware, but there are a few devices with RTL8221B phys in the realtek target.  Naturally, I guess :-)

Don't know if any of them are currently supported, though.  I don't have any.  And they are quite expensive toys.,  But it is probably wise to co-ordinate any work.  Here's one (unsupported) device exapmle: https://svanheule.net/switches/ms510tx

NOTE: only the TXM or TXUP variants! Netgear don't have enough model numbers to uniquely name different platforms

I believe there are some others as well.  Maybe ask in the nice, short https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875 thread?

ptpt52 commented:
> 721-net-phy-realtek-rtl8221-allow-to-configure-SERDES-mo.patch

@dangowrt 
sure I build the firmware with this patch [721-net-phy-realtek-rtl8221-allow-to-configure-SERDES-mo.patch](https://github.com/x-wrt/x-wrt/blob/master/target/linux/generic/pending-5.15/721-net-phy-realtek-rtl8221-allow-to-configure-SERDES-mo.patch)
it do not resolve any of the issues:
1. 2.5G phy connected to first mt7531 P5 link up ok, but cannot send/recv data
2. 2.5G phy connected to 2nd-mt7531 cannot even detect link.

I am sure the on the SDK code, all 2.5G phy link up ok, and send/recv data ok.

dangowrt commented:
> but the P5 still have to be poll mode, what to do to add interrupt support for it?
> looks like we have to add code to support it in `drivers/net/dsa/mt7530.c` ?

No. In case the PHY has the irq line connected you can just wire it up by setting `interrupt-parent` and `interrupts` in DTS for that PHY. In any case, not having to poll all the built-in PHYs of the MT7531 should already significantly relax the load on the bit-banged gpio-driven MDIO bus.

ptpt52 commented:
> > but the P5 still have to be poll mode, what to do to add interrupt support for it?
> > looks like we have to add code to support it in `drivers/net/dsa/mt7530.c` ?
> 
> No. In case the PHY has the irq line connected you can just wire it up by setting `interrupt-parent` and `interrupts` in DTS for that PHY. In any case, not having to poll all the built-in PHYs of the MT7531 should already significantly relax the load on the bit-banged gpio-driven MDIO bus.

somewhat like this?
```diff
&mdio_bus {
        phy7: ethernet-phy@7 {
                reg = <7>;
                compatible = "ethernet-phy-ieee802.3-c45";
                phy-mode = "2500base-x";
                reset-gpios = <&pio 14 GPIO_ACTIVE_LOW>;
                reset-delay-us = <600>;
                reset-post-delay-us = <20000>;
                interrupt-parent = <&pio>;
                interrupts = <7 IRQ_TYPE_LEVEL_HIGH>; //if phy init connect to soc gpio7
        };
```

but I cannot find any example of this. and no kernel code to handle this
```
interrupt-parent = <&pio>;
 interrupts = <7 IRQ_TYPE_LEVEL_HIGH>;
```

dangowrt commented:
> and no kernel code to handle this

Kernel does handle this and almost all PHY drivers support handling interrupts instead of polling:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/net/phy/phy.c#n1274

This is true also for RTL8221B, see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/net/phy/realtek.c#n181

> ```
> interrupt-parent = <&pio>;
>  interrupts = <7 IRQ_TYPE_LEVEL_HIGH>;
> ```
This looks good.
